File: 214-shortest-palindrome/shortest-palindrome.cpp
class Solution {
public:
    string shortestPalindrome(string s) {
        int n = s.length();
        if (n == 0) return s;
        int prefix = 0;
        for (int i = n - 1; i >= 0; i--) {
            if (is_palindrome(s, 0, i)) {
                prefix = i + 1;
                break;
            }
        }
        string suffix = s.substr(prefix);
        reverse(suffix.begin(), suffix.end());
        return suffix + s;
    } 
    bool is_palindrome(string& s, int start, int end) {
        while (start < end) {
            if (s[start] != s[end]) return false;
            start++;
            end--;
        }
        return true;
    }
};

File: 214-shortest-palindrome/shortest-palindrome.java
class Solution {
    public String shortestPalindrome(String s) {
        String rev_s = new StringBuilder(s).reverse().toString();
        String combined = s + "#" + rev_s;
        int[] lps = new int[combined.length()];
        for (int i = 1; i < combined.length(); i++) {
            int j = lps[i - 1]; 
            while (j > 0 && combined.charAt(i) != combined.charAt(j)) {
                j = lps[j - 1];
            }
            if (combined.charAt(i) == combined.charAt(j)) {
                j++;
            }
            lps[i] = j;
        }
        int longest_palindrome_len = lps[combined.length() - 1];
        String to_add = rev_s.substring(0, s.length() - longest_palindrome_len);
        return to_add + s;
    }
}
